be/test/storage/lake/local_pk_index_manager_test.cpp

```diff
@@ -173,6 +173,110 @@ TEST_F(LocalPkIndexManagerTest, test_gc) {
+    auto stores = StorageEngine::instance()->get_stores();
+    ASSERT_TRUE(stores.size() > 0);
+    ASSERT_OK(FileSystem::Default()->path_exists(stores[0]->get_persistent_index_path() + "/" +
+                                                 std::to_string(_tablet_metadata->id())));
+
+    auto* data_dir = stores[0];
+    auto pk_path = data_dir->get_persistent_index_path();
+    std::set<std::string> tablet_ids;
+    ASSERT_OK(fs::list_dirs_files(pk_path, &tablet_ids, nullptr));
+
+    {
+        // lock acquire failed
+        SyncPoint::GetInstance()->SetCallBack("LocalPkIndexManager:index_dir_changed:3",
+                                              [](void* arg) { *(bool*)arg = false; });
+        LocalPkIndexManager::gc(ExecEnv::GetInstance()->lake_update_manager(), data_dir, tablet_ids);
+        // no index dir removed
+        ASSERT_OK(FileSystem::Default()->path_exists(stores[0]->get_persistent_index_path() + "/" +
+                                                     std::to_string(_tablet_metadata->id())));
+        SyncPoint::GetInstance()->ClearCallBack("LocalPkIndexManager:index_dir_changed:3");
+    }
+
+    {
+        // lock acquire succeed
+        SyncPoint::GetInstance()->SetCallBack("LocalPkIndexManager:index_dir_changed:3",
+                                              [](void* arg) { *(bool*)arg = true; });
+        LocalPkIndexManager::gc(ExecEnv::GetInstance()->lake_update_manager(), data_dir, tablet_ids);
+        // index dir removed
+        ASSERT_ERROR(FileSystem::Default()->path_exists(stores[0]->get_persistent_index_path() + "/" +
+                                                        std::to_string(_tablet_metadata->id())));
+        SyncPoint::GetInstance()->ClearCallBack("LocalPkIndexManager:index_dir_changed:3");
+    }
+
+    SyncPoint::GetInstance()->ClearCallBack("LocalPkIndexManager:index_dir_changed:1");
+    SyncPoint::GetInstance()->ClearCallBack("LocalPkIndexManager:index_dir_changed:2");
+    SyncPoint::GetInstance()->DisableProcessing();
+
+    txn_id = next_id();
+    ASSIGN_OR_ABORT(writer, tablet.new_writer(kHorizontal, txn_id));
+    ASSERT_OK(writer->open());
+
+    // write segment #2
+    ASSERT_OK(writer->write(chunk0));
+    ASSERT_OK(writer->finish());
+
+    // write txnlog
+    txn_log = std::make_shared<TxnLog>();
+    txn_log->set_tablet_id(_tablet_metadata->id());
+    txn_log->set_txn_id(txn_id);
+    op_write = txn_log->mutable_op_write();
+    for (auto& f : writer->files()) {
+        op_write->mutable_rowset()->add_segments(std::move(f.path));
+    }
+    op_write->mutable_rowset()->set_num_rows(writer->num_rows());
+    op_write->mutable_rowset()->set_data_size(writer->data_size());
+    ASSERT_OK(_tablet_mgr->put_txn_log(txn_log));
+
+    writer->close();
+    // publish again should be successful, persistent index will be reloaded.
+    ASSERT_OK(publish_single_version(_tablet_metadata->id(), 2, txn_id).status());
+}
```

Comment:
looks like the ut doesn't cover the case that multiple data_dirs exists for the same index which is one of the scenario this PR is trying to fix. right?

Reply:
used `LocalPkIndexManager:index_dir_changed:2` to mock this case, maybe it can cover partly..